node/beacon-chain-handler/src/importer.rs

```diff
@@ -58,69 +67,98 @@ impl BeaconBlockImporter {
         true
     }
 
+    pub fn import_shard_block(&mut self, shard_block: ShardBlock) {
+        let hash = shard_block.hash();
+        if !self.pending_shard_blocks.contains_key(&hash) {
+            self.pending_shard_blocks.insert(hash, shard_block);
+        }
+    }
+
+    fn add_block(&self, block: BeaconBlock, shard_block: ShardBlock) {
+        let parent_hash = block.body.header.parent_hash;
+        let parent_shard_hash = shard_block.body.header.parent_hash;
+        let num_transactions = shard_block.body.transactions.len();
+        let num_receipts = shard_block.body.receipts.len();
+        // we can unwrap because parent is guaranteed to exist
+        let prev_header = self.beacon_chain
+            .get_header(&BlockId::Hash(parent_hash))
+            .expect("Parent is known but header not found.");
+        let prev_shard_header = self.shard_chain
+            .get_header(&BlockId::Hash(parent_shard_hash))
+            .expect("At this moment shard chain should be present together with beacon chain");
+        let apply_state = ApplyState {
+            root: prev_shard_header.body.merkle_root_state,
+            block_index: prev_header.body.index,
+            parent_block_hash: parent_hash,
+        };
+        let mut receipts = shard_block.body.receipts.clone();
+        let (filtered_transactions, filtered_receipts, mut apply_result) =
+            self.runtime.write().apply(&apply_state, shard_block.body.transactions.clone(), &mut receipts);
+        if apply_result.root != prev_shard_header.body.merkle_root_state {
```

Comment:
do we want to fail silently for the following 3 cases for now?

Reply:
Discussed offline, should add later some kind of repoter/aggregator for cases like this for further reaction.